Enable Native Markdown Extraction in Apache PDFBox

rawatsaurav01 opened the pull request:
Propose the addition of native Markdown extraction support in Apache PDFBox to simplify the conversion of PDF content to Markdown, eliminating the need for intermediate HTML conversion.

**Description:**
Currently, Apache PDFBox supports HTML extraction through `PdfText2HTML`. However, this requires an extra step of converting HTML to Markdown using external tools like CopyDown. To enhance efficiency, we suggest incorporating native Markdown extraction support within Apache PDFBox.

**Sample Code Comparison:**

**Current Process:**

```java
File pdfFile = new File("sample/sample.pdf");
File mdFile = new File("sample/sample.md");
PDFText2HTML pdfText2HTML = new PDFText2HTML();
CopyDown copyDown = new CopyDown();
try (PDDocument pdDocument = Loader.loadPDF(pdfFile)) {
    Files.writeString(mdFile.toPath(), copyDown.convert(pdfText2HTML.getText(pdDocument)));
}

**Proposed Process:**

```java
File pdfFile = new File("sample/sample.pdf");
File mdFile = new File("sample/sample.md");
PDFText2Markdown pdfText2Markdown = new PDFText2Markdown();
try (PDDocument pdDocument = Loader.loadPDF(pdfFile)) {
    Files.writeString(mdFile.toPath(), pdfText2Markdown.getText(pdDocument));
}

**Benefits:**

1. **Streamlined Workflow:** Direct PDF to Markdown conversion without relying on external tools.
3. **Performance Improvement:** Reduced resource consumption, especially for large PDF files.
4. **Enhanced User Experience:** Aligns with common use cases, improving overall usability.

**Proposed Changes:**

Introduce `PDFText2Markdown` in Apache PDFBox to provide native Markdown extraction.

**Compatibility:**

Ensure backward compatibility with existing PDFBox functionalities while seamlessly adding Markdown extraction.

rawatsaurav01 committed:
added PDFText2Markdown class

rawatsaurav01 committed:
modified PDFText2Markdown class

THausherr commented:
Seems like a good idea. @lehmi do we need an ICLA for this one?

lehmi commented:
@THausherr IMHO it is not needed as the class is more or less a copy of PDFText2HTML

lehmi commented on tools/src/main/java/org/apache/pdfbox/tools/PDFText2Markdown.java:
```diff
@@ -0,0 +1,375 @@
+        StringBuilder buf = new StringBuilder(INITIAL_PDF_TO_HTML_BYTES);
+
+        super.writeString(buf.toString());
+    }
```
IMHO, this doesn't make sense. The used StringBuilder is empty so that nothing is written at all.

rawatsaurav01 replied:
@lehmi I have removed line as per the suggestion.

lehmi commented on tools/src/main/java/org/apache/pdfbox/tools/PDFText2Markdown.java:
```diff
@@ -0,0 +1,375 @@
+            super.writeString(LINE_SEPARATOR);
+        }
+    }
+
```
Both cases are doing the same, so that the if/else can be removed

rawatsaurav01 replied:
@lehmi Please check I have updated the source code.

lehmi commented:
@rawatsaurav01 thanks for the contribution.  I've added some comments to your code and here are some more

* add a apache license header to the file
* remove/replace all HTML related comments/javadocs
* remove commented code

rawatsaurav01 committed:
implemented code review comments

rawatsaurav01 commented:
@lehmi @THausherr Thanks a lot for your encouraging comments . I have encorporated the changes. Please review and let me know if some changes required.

THausherr reviewed (changes requested):
Please put back Axels name at line 164, he should still be credited as your co-author IMHO.

rawatsaurav01 committed:
implemented code review comments-2

rawatsaurav01 commented:
@THausherr I have changed the author name to Axle . Can you please check it now.

THausherr commented:
Thank you @rawatsaurav01 I committed to the trunk with some minor changes. I'll commit to 3.0 and 2.0 later.

rawatsaurav01 commented:
Thank you @THausherr .

THausherr commented:
Sonar found a missing break in appendEscaped() which is no big deal, but then I realized that some of the code isn't needed. At the same time some needed code should be added, i.e. handling of special characters like "*". See also 
https://shadocs.netlify.app/markdown/escaping_character/

rawatsaurav01 commented:
Hi @THausherr I have done some changes in the code. Previously it was having issue in handling special characters like *,#,+,-. But now the issue is resolved . Actually it takes * as dot and # as a line so in the new edit I have added  a backslash('\') for the characters which are giving errors.
 I have added a screenshot of https://markdownlivepreview.com/ after converting a pdf (having special charaters) to markdown. 
![Screenshot 2024-02-17 232046](https://github.com/apache/pdfbox/assets/90169524/4b3ea6fd-224c-4ba5-9275-d2baa1460051)

THausherr commented:
I don't see any additional commit, either add it or post the code change here. I assume it's in `appendEscaped()`.

rawatsaurav01 commented:
@THausherr I have added and pushed the code ,but it is not showing maybe the pr is closed.Here is the updated code.

```java
private static void appendEscaped(StringBuilder builder, char character) {

        switch (character) {
            case '*':
                builder.append("\\" + character);
                break;
            case '+':
                builder.append("\\" + character);
                break;
            case '-':
                builder.append("\\" + character);
                break;
            case '#':
                builder.append("\\" + character);
                break;
            case 34:
                builder.append("\"");
                break;
            case 38:
                builder.append("&");
                break;
            case 39:
                builder.append("\'");
                break;
            case 60:
                builder.append("<");
                break;
            case 62:
                builder.append(">");
                break;
            case 178:
                builder.append("<sup>2</sup>");
                break;
            case 179:
                builder.append("<sup>3</sup>");
                break;

            default:
                builder.append(character);
                break;
        }
    }

```

THausherr commented:
Thank you, I've simplified this and committed it. If there isn't any further input then I'll commit to versions 2 and 3 in a few days.

rawatsaurav01 commented:
Thank you @THausherr for your acknowledgement.

THausherr commented:
All done in https://issues.apache.org/jira/browse/PDFBOX-5768 .

johmue commented:
@THausherr @rawatsaurav01 This is really helpful, especially now with LLMs trained on MarkDown formatted documents. Would it be a big effort to expose this option in the PDFBox App with a dedicated SubCommand i.e. `export:md` or with `export:text -md` (similar to -html)?

THausherr commented:
Yes sounds like a good idea, I'm gonna think about it tomorrow.

THausherr commented:
Done  https://issues.apache.org/jira/browse/PDFBOX-5895 .
It will soon be available at
https://repository.apache.org/content/groups/snapshots/org/apache/pdfbox/pdfbox-app/2.0.33-SNAPSHOT/
https://repository.apache.org/content/groups/snapshots/org/apache/pdfbox/pdfbox-app/3.0.4-SNAPSHOT/
Please test it.

johmue commented:
It works and creates good outcomes. Thanks for putting the effort and making it easily available in almost no time.
